modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/GeoIpDatabase.java

```diff
@@ -34,44 +30,18 @@ public interface GeoIpDatabase {
     String getDatabaseType() throws IOException;
 
     /**
-     * @param ipAddress the IP address to look up
-     * @return a response containing the city data for the given address if it exists, or <code>null</code> if it could not be found
-     * @throws UnsupportedOperationException may be thrown if the implementation does not support retrieving city data
+     * Returns a response from this database's reader for the given IP address
+     * @param ipAddress the address to lookup
+     * @param responseProvider typically a method-reference like {@code DatabaseReader::tryCity}
+     * @return a possibly-null response
+     * @param <RESPONSE> the subtype of {@link AbstractResponse} that will be returned
+     * @throws IOException if the implementation encounters any problem while retrieving the response
      */
     @Nullable
-    CityResponse getCity(InetAddress ipAddress);
-
-    /**
-     * @param ipAddress the IP address to look up
-     * @return a response containing the country data for the given address if it exists, or <code>null</code> if it could not be found
-     * @throws UnsupportedOperationException may be thrown if the implementation does not support retrieving country data
-     */
-    @Nullable
-    CountryResponse getCountry(InetAddress ipAddress);
-
-    /**
-     * @param ipAddress the IP address to look up
-     * @return a response containing the Autonomous System Number for the given address if it exists, or <code>null</code> if it could not
-     *         be found
-     * @throws UnsupportedOperationException may be thrown if the implementation does not support retrieving ASN data
-     */
-    @Nullable
-    AsnResponse getAsn(InetAddress ipAddress);
-
-    @Nullable
-    AnonymousIpResponse getAnonymousIp(InetAddress ipAddress);
-
-    @Nullable
-    ConnectionTypeResponse getConnectionType(InetAddress ipAddress);
-
-    @Nullable
-    DomainResponse getDomain(InetAddress ipAddress);
-
-    @Nullable
-    EnterpriseResponse getEnterprise(InetAddress ipAddress);
-
-    @Nullable
-    IspResponse getIsp(InetAddress ipAddress);
+    <RESPONSE extends AbstractResponse> RESPONSE getResponse(
```

Comment:
The `GeoIpDatabase` interface is still tightly-coupled with maxmind's `DatabaseReader`, but it is no longer tightly-coupled with the _types_ of responses we can get from it, so implementations of this interface will no longer need to be modified when we add support for new types here.

Reply:
There is an opportunity here to use [`<T> com.maxmind.db.Reader#get(InetAddress, Class<T>)`](https://maxmind.github.io/MaxMind-DB-Reader-java/doc/v3.1.0/com.maxmind.db/com/maxmind/db/Reader.html) _directly_ instead of proxying through to the model-bound `com.maxmind.geoip2.DatabaseReader.tryXXX(InetAddress)` methods, which would allow for the use of mmdb's whose contents do not adhere to one of the geoip2 models.

This would obviously be a larger refactor that would reach into the node service's Geoip provider that currently provides `com.maxmind.geoip2.DatabaseReader`, but may be required if we ever support mmdb's from other vendors.